### Problem Statement:

ABC Bank wants to sell it's term deposit product to customers and before launching the product they want to develop a model which help them in understanding whether a particular customer will buy their product or not (based on customer's past interaction with bank or other Financial Institution).

 

#### Why ML Model: 
Bank wants to use ML model to shortlist customer whose chances of buying the product is more so that their marketing channel (tele marketing, SMS/email marketing etc)  can focus only to those customers whose chances of buying the product is more.

This will save resource and their time ( which is directly involved in the cost ( resource billing)).

Develop model with Duration and without duration feature and report the performance of the model.

Duration feature is not recommended as this will be difficult to explain the result to business and also it will

be difficult for business to campaign based on duration.

Imbalance data set : If dataset is imbalance then try to use techniques to handle imbalance data set and report model performance of each imbalance technique used (avoid techniques which recommend not to utilize whole available data)

#### Task:

Business Understanding
Data understanding
Exploratory data Analysis
Data Preparation
Model Building ( Logistic Regression, ensemble, Boosting etc)
Model Selection
Performance reporting
Deploy the model
Converting ML metrics into Business metric and explaining result to business
Prepare presentation for non technical persons.


In [1]:
import numpy as np
import pandas as pd

In [2]:
dt = pd.read_csv('bank-additional-full.csv', sep=';')

In [3]:
dt.head()

age        job  marital    education  default housing loan    contact  \
0   56  housemaid  married     basic.4y       no      no   no  telephone   
1   57   services  married  high.school  unknown      no   no  telephone   
2   37   services  married  high.school       no     yes   no  telephone   
3   40     admin.  married     basic.6y       no      no   no  telephone   
4   56   services  married  high.school       no      no  yes  telephone   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         0  nonexistent          1.1   
1   may         mon  ...         1    999         0  nonexistent          1.1   
2   may         mon  ...         1    999         0  nonexistent          1.1   
3   may         mon  ...         1    999         0  nonexistent          1.1   
4   may         mon  ...         1    999         0  nonexistent          1.1   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          93.994          -36.4      4.857       5191.0  no  
1          93.994          -36.4      4.857       5191.0  no  
2          93.994          -36.4      4.857       5191.0  no  
3          93.994          -36.4      4.857       5191.0  no  
4          93.994          -36.4      4.857       5191.0  no  

[5 rows x 21 columns]

In [4]:
dt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41188 entries, 0 to 41187
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   age             41188 non-null  int64  
 1   job             41188 non-null  object 
 2   marital         41188 non-null  object 
 3   education       41188 non-null  object 
 4   default         41188 non-null  object 
 5   housing         41188 non-null  object 
 6   loan            41188 non-null  object 
 7   contact         41188 non-null  object 
 8   month           41188 non-null  object 
 9   day_of_week     41188 non-null  object 
 10  duration        41188 non-null  int64  
 11  campaign        41188 non-null  int64  
 12  pdays           41188 non-null  int64  
 13  previous        41188 non-null  int64  
 14  poutcome        41188 non-null  object 
 15  emp.var.rate    41188 non-null  float64
 16  cons.price.idx  41188 non-null  float64
 17  cons.conf.idx   41188 non-null 

   ### Input variables:
   
   #### bank client data:
   
   1 - age (numeric)
   
   2 - job : type of job 
   
   categorical: "admin.", "blue-collar", "entrepreneur", "housemaid", "management", "retired",
                "self-employed", "services", "student", "technician", "unemployed", "unknown"
                
   3 - marital : marital status 
   
   categorical: "divorced", "married", "single", "unknown"; 
   (note: "divorced" means divorced or widowed)
   
   4 - education 
   
   categorical: "basic.4y", "basic.6y", "basic.9y", "high.school", "illiterate", 
                "professional.course", "university.degree", "unknown"
                
   5 - default: has credit in default? (categorical: "no","yes","unknown")
   
   6 - housing: has housing loan? (categorical: "no","yes","unknown")
   
   7 - loan: has personal loan? (categorical: "no","yes","unknown")
   
   #### related with the last contact of the current campaign:
   
   8 - contact: contact communication type (categorical: "cellular","telephone") 
   
   9 - month: last contact month of year (categorical: "jan", "feb", "mar", ..., "nov", "dec")
   
  10 - day_of_week: last contact day of the week (categorical: "mon", "tue", "wed", "thu", "fri")
  
  11 - duration: last contact duration, in seconds (numeric). 
  
  Important note:  this attribute highly affects the output target (e.g., if duration=0 then y="no"). 
  
  Yet, the duration is not known before a call is performed. 
  
  Also, after the end of the call y is obviously known. 
  
  Thus, this input should only be included for benchmark purposes and should be discarded if the intention is to have a realistic predictive model.
  
   #### other attributes:
   
  12 - campaign: number of contacts performed during this campaign and for this client 
  (numeric, includes last contact)
  
  13 - pdays: number of days that passed by after the client was last contacted from a previous campaign 
  (numeric; 999 means client was not previously contacted)
  
  14 - previous: number of contacts performed before this campaign and for this client (numeric)
  
  15 - poutcome: outcome of the previous marketing campaign 
  (categorical: "failure", "nonexistent", "success")
  
   #### social and economic context attributes
   
  16 - emp.var.rate: employment variation rate - quarterly indicator (numeric)
  
  17 - cons.price.idx: consumer price index - monthly indicator (numeric)     
  
  18 - cons.conf.idx: consumer confidence index - monthly indicator (numeric)   
  
  19 - euribor3m: euribor 3 month rate - daily indicator (numeric)
  
  20 - nr.employed: number of employees - quarterly indicator (numeric)

  ### Output variable (desired target):
  
  21 - y - has the client subscribed a term deposit? (binary: "yes","no")

In [5]:
dt['job'].value_counts()

job
admin.           10422
blue-collar       9254
technician        6743
services          3969
management        2924
retired           1720
entrepreneur      1456
self-employed     1421
housemaid         1060
unemployed        1014
student            875
unknown            330
Name: count, dtype: int64

In [6]:
dt['marital'].value_counts()

marital
married     24928
single      11568
divorced     4612
unknown        80
Name: count, dtype: int64

In [7]:
dt['education'].value_counts()

education
university.degree      12168
high.school             9515
basic.9y                6045
professional.course     5243
basic.4y                4176
basic.6y                2292
unknown                 1731
illiterate                18
Name: count, dtype: int64

In [8]:
dt['default'].value_counts()

default
no         32588
unknown     8597
yes            3
Name: count, dtype: int64

In [9]:
dt['housing'].value_counts()

housing
yes        21576
no         18622
unknown      990
Name: count, dtype: int64

In [10]:
dt['loan'].value_counts()

loan
no         33950
yes         6248
unknown      990
Name: count, dtype: int64

In [11]:
dt['contact'].value_counts()

contact
cellular     26144
telephone    15044
Name: count, dtype: int64

In [12]:
dt['month'].value_counts()

month
may    13769
jul     7174
aug     6178
jun     5318
nov     4101
apr     2632
oct      718
sep      570
mar      546
dec      182
Name: count, dtype: int64

In [13]:
dt['day_of_week'].value_counts()

day_of_week
thu    8623
mon    8514
wed    8134
tue    8090
fri    7827
Name: count, dtype: int64

In [14]:
dt['poutcome'].value_counts()

poutcome
nonexistent    35563
failure         4252
success         1373
Name: count, dtype: int64

In [15]:
dt['y'].value_counts()

y
no     36548
yes     4640
Name: count, dtype: int64

### The dataset reveals the presence of missing data in six categorical features: job, education, marital status, default, housing, and loan.

### The dataset displays an imbalance, particularly in the target variable for the predictive classification model, where 90% of the cases are skewed towards the "N" category.

In [16]:
dt.describe()

age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      1.000000    999.000000      0.000000   
50%       38.00000    180.000000      2.000000    999.000000      0.000000   
75%       47.00000    319.000000      3.000000    999.000000      0.000000   
max       98.00000   4918.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  41188.000000    41188.000000   41188.000000  41188.000000  41188.000000  
mean       0.081886       93.575664     -40.502600      3.621291   5167.035911  
std        1.570960        0.578840       4.628198      1.734447     72.251528  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

### There is one numerical feature, "duration," exhibiting outlier data. Notably, the mean duration is approximately 258, but the maximum value is 4918, signaling the presence of outliers.

## Data Cleaning and Transformation

### Dropping missing values in job and marital

In [17]:
dt[(dt['marital'] == 'unknown')]

age           job  marital          education  default housing loan  \
40      58    management  unknown  university.degree       no     yes   no   
390     59       retired  unknown  university.degree  unknown      no   no   
413     57       retired  unknown           basic.4y       no      no   no   
1493    33      services  unknown        high.school       no     yes  yes   
1608    32       unknown  unknown  university.degree       no      no   no   
...    ...           ...      ...                ...      ...     ...  ...   
38550   31  entrepreneur  unknown  university.degree       no      no   no   
38579   31  entrepreneur  unknown  university.degree       no     yes   no   
39322   30    technician  unknown  university.degree       no      no   no   
39323   30    technician  unknown  university.degree       no      no   no   
40857   30    technician  unknown  university.degree       no      no   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
40     telephone   may         mon  ...         1    999         0   
390    telephone   may         tue  ...         1    999         0   
413    telephone   may         tue  ...         1    999         0   
1493   telephone   may         thu  ...         2    999         0   
1608   telephone   may         fri  ...         4    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
38550   cellular   oct         thu  ...         1    999         0   
38579  telephone   oct         thu  ...         4    999         1   
39322   cellular   mar         mon  ...         2     11         2   
39323   cellular   mar         mon  ...         2      6         2   
40857   cellular   oct         fri  ...         2     16         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
40     nonexistent          1.1          93.994          -36.4      4.857   
390    nonexistent          1.1          93.994          -36.4      4.857   
413    nonexistent          1.1          93.994          -36.4      4.857   
1493   nonexistent          1.1          93.994          -36.4      4.855   
1608   nonexistent          1.1          93.994          -36.4      4.855   
...            ...          ...             ...            ...        ...   
38550  nonexistent         -3.4          92.431          -26.9      0.722   
38579      failure         -3.4          92.431          -26.9      0.722   
39322      success         -1.8          93.369          -34.8      0.639   
39323      success         -1.8          93.369          -34.8      0.639   
40857      success         -1.1          94.601          -49.5      0.942   

       nr.employed    y  
40          5191.0   no  
390         5191.0   no  
413         5191.0   no  
1493        5191.0   no  
1608        5191.0  yes  
...            ...  ...  
38550       5017.5  yes  
38579       5017.5   no  
39322       5008.7  yes  
39323       5008.7  yes  
40857       4963.6   no  

[80 rows x 21 columns]

In [18]:
drop_marital = dt[(dt['marital'] == 'unknown')].index
dt.drop(drop_marital, inplace=True)

In [19]:
dt[(dt['job'] == 'unknown')]

age      job   marital            education  default  housing     loan  \
29      55  unknown   married    university.degree  unknown  unknown  unknown   
35      55  unknown   married             basic.4y  unknown      yes       no   
73      57  unknown   married              unknown  unknown       no       no   
91      57  unknown   married              unknown  unknown      yes       no   
144     38  unknown  divorced          high.school  unknown      yes       no   
...    ...      ...       ...                  ...      ...      ...      ...   
40370   59  unknown   married              unknown       no       no       no   
40428   64  unknown   married              unknown       no      yes       no   
40656   67  unknown  divorced              unknown  unknown      yes       no   
41005   63  unknown   married  professional.course       no       no       no   
41108   64  unknown   married              unknown       no       no       no   

         contact month day_of_week  ...  campaign  pdays  previous  \
29     telephone   may         mon  ...         1    999         0   
35     telephone   may         mon  ...         1    999         0   
73     telephone   may         mon  ...         1    999         0   
91     telephone   may         mon  ...         1    999         0   
144    telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
40370   cellular   aug         wed  ...         1      6         1   
40428  telephone   aug         fri  ...         4    999         0   
40656   cellular   sep         wed  ...         2      6         2   
41005   cellular   oct         thu  ...         1      6         1   
41108   cellular   nov         mon  ...         1      3         3   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
29     nonexistent          1.1          93.994          -36.4      4.857   
35     nonexistent          1.1          93.994          -36.4      4.857   
73     nonexistent          1.1          93.994          -36.4      4.857   
91     nonexistent          1.1          93.994          -36.4      4.857   
144    nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
40370      success         -1.7          94.027          -38.3      0.900   
40428  nonexistent         -1.7          94.027          -38.3      0.905   
40656      success         -1.1          94.199          -37.5      0.880   
41005      success         -1.1          94.601          -49.5      1.025   
41108      success         -1.1          94.767          -50.8      1.048   

       nr.employed    y  
29          5191.0   no  
35          5191.0   no  
73          5191.0   no  
91          5191.0   no  
144         5191.0   no  
...            ...  ...  
40370       4991.6  yes  
40428       4991.6  yes  
40656       4963.6  yes  
41005       4963.6   no  
41108       4963.6   no  

[321 rows x 21 columns]

In [20]:
drop_job = dt[(dt['job'] == 'unknown')].index
dt.drop(drop_job, inplace=True)

In [21]:
updated = dt['job'] == 'admin.'
dt.loc[updated, 'job'] = 'admin'
dt

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40        admin  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[40787 rows x 21 columns]

In [22]:
dt['job'].value_counts()

job
admin            10408
blue-collar       9240
technician        6731
services          3963
management        2921
retired           1715
entrepreneur      1453
self-employed     1416
housemaid         1057
unemployed        1009
student            874
Name: count, dtype: int64

In [23]:
dt['marital'].value_counts()

marital
married     24694
single      11494
divorced     4599
Name: count, dtype: int64

### Replacing missing values in housing and loan with mode

In [24]:
mode_housing = dt['housing'].mode()[0]
updated_housing = dt['housing'] == 'unknown'
dt.loc[updated_housing, 'housing'] = mode_housing

mode_loan = dt['loan'].mode()[0]
updated_loan = dt['loan'] == 'unknown'
dt.loc[updated_loan, 'loan'] = mode_loan

dt

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40        admin  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[40787 rows x 21 columns]

In [25]:
dt['housing'].value_counts()

housing
yes    22360
no     18427
Name: count, dtype: int64

In [26]:
dt['loan'].value_counts()

loan
no     34604
yes     6183
Name: count, dtype: int64

### Fixing outliers in duration column

In [27]:
#Computing the upper bound for the duration column
Q3 = dt['duration'].quantile(0.75)
IQR = dt['duration'].quantile(0.75) - dt['duration'].quantile(0.25)
upper_bound = Q3 + 3*IQR

# Remove outliers
dt[dt['duration'] <= upper_bound]
dt = dt[dt['duration'] <= upper_bound]
dt

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40        admin  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[39758 rows x 21 columns]

In [28]:
dt.describe()

age      duration      campaign         pdays      previous  \
count  39758.000000  39758.000000  39758.000000  39758.000000  39758.000000   
mean      39.991247    230.762438      2.565069    962.685472      0.174003   
std       10.410769    185.284932      2.779073    186.392779      0.496998   
min       17.000000      0.000000      1.000000      0.000000      0.000000   
25%       32.000000    101.000000      1.000000    999.000000      0.000000   
50%       38.000000    175.000000      2.000000    999.000000      0.000000   
75%       47.000000    303.000000      3.000000    999.000000      0.000000   
max       98.000000    972.000000     56.000000    999.000000      7.000000   

       emp.var.rate  cons.price.idx  cons.conf.idx     euribor3m   nr.employed  
count  39758.000000    39758.000000   39758.000000  39758.000000  39758.000000  
mean       0.078346       93.573942     -40.509907      3.617127   5166.911573  
std        1.570663        0.578767       4.629099      1.735227     72.237558  
min       -3.400000       92.201000     -50.800000      0.634000   4963.600000  
25%       -1.800000       93.075000     -42.700000      1.344000   5099.100000  
50%        1.100000       93.749000     -41.800000      4.857000   5191.000000  
75%        1.400000       93.994000     -36.400000      4.961000   5228.100000  
max        1.400000       94.767000     -26.900000      5.045000   5228.100000

### There are missing values in the education and default columns

In [29]:
dt['education'].value_counts()

education
university.degree      11793
high.school             9226
basic.9y                5846
professional.course     5113
basic.4y                4017
basic.6y                2194
unknown                 1552
illiterate                17
Name: count, dtype: int64

In [30]:
dt['default'].value_counts()

default
no         31519
unknown     8236
yes            3
Name: count, dtype: int64

### Let's leave the value 'unknown' in these columns. Here, it can be justified.